### PR TITLE
HPCC-13331 Should be able to despray XML without rowTag specification.

### PR DESCRIPTION
**Pull request by AttilaVamos**
Add code to store xml header/footer lenght into the file part metadata on
hthor, thor and roxie when:
- spray xml file or files (with different xml header/footer)
- output data in XML format w/wo custom header/footer/ row tag

Add code to use stored header/footer lenght in despray XML data

Signed-off-by: Attila Vamos attila.vamos@gmail.com

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith Please review.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

File: dali/ft/filecopy.cpp [hunk 1]
```diff
@@ -1054,7 +1056,7 @@ void FileSprayer::calculateMany2OnePartition()
         setCanAccessDirectly(curFilename);
         if (partSeparator)
         {
-            offset_t contentLength = cur.size - cur.xmlHeaderLength - cur.xmlFooterLength;
+            offset_t contentLength = (cur.size > cur.xmlHeaderLength + cur.xmlFooterLength ? cur.size - cur.xmlHeaderLength - cur.xmlFooterLength : 0);
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: (cur.size > cur.xmlHeaderLength + cur.xmlFooterLength ? cur.size - cur.xmlHeaderLength - cur.xmlFooterLength : 0);

Can cur.size be > 0 , but < (xmlHeaderLength+xmlFooterLength) ?
e.g. when no header/footer.

If so, looks like this will incorrectly set contentLength to 0.

File: dali/ft/filecopy.cpp [hunk 2]
```diff
@@ -1557,12 +1569,36 @@ void FileSprayer::locateXmlHeader(IFileIO * io, unsigned headerSize, offset_t &
...
+    {
+        try
+        {
+            xmlHeaderLength = splitter.getHeaderLength(reader);
+        }
+        catch (IException * e)
+        {
+            if (e->errorCode() != DFTERR_CannotFindFirstXmlRecord)
+                throw;
+            xmlHeaderLength = 0;
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: Leaks 'e'

File: dali/ft/filecopy.cpp [hunk 3]
```diff
@@ -1557,12 +1569,36 @@ void FileSprayer::locateXmlHeader(IFileIO * io, unsigned headerSize, offset_t &
...
+    {
+        try
+        {
+            xmlFooterLength = splitter.getFooterLength(reader, size);
+        }
+        catch (IException * e)
+        {
+            if (e->errorCode() != DFTERR_CannotFindLastXmlRecord)
+                throw;
+            xmlFooterLength= 0;
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: Leaks 'e'

File: dali/ft/filecopy.cpp [hunk 4]
```diff
@@ -1945,7 +1981,7 @@ void FileSprayer::cloneHeaderFooter(unsigned idx, bool isHeader)
     if (isHeader)
         next.inputOffset = curSrc.headerSize - curSrc.xmlHeaderLength;
     else
-        next.inputOffset = curSrc.headerSize + curSrc.size;
+        next.inputOffset = curSrc.size + curSrc.xmlHeaderLength;
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: I'm a bit confused by this change.. may need a comment..
why in isHeader=false (so footer?) .. is it adding xmlHeaderLength here?

File: dali/ft/filecopy.cpp [hunk 5]
```diff
@@ -2873,6 +2910,16 @@ void FileSprayer::updateTargetProperties()
...
                 curPart->lockProperties();
                 IPropertyTree& curProps = curPart->queryAttributes();
+
+                FilePartInfo & curHeaderSource = sources.item(whichHeaderInput);
+                curProps.setPropInt(FAheaderLength, curHeaderSource.xmlHeaderLength);
+
+                FilePartInfo & curFooterSource = sources.item(cur.whichInput);
+                curProps.setPropInt(FAfooterLength, curFooterSource.xmlFooterLength);
+
+                if( idx+1 != partition.ordinality() && partition.item(idx+1).whichOutput != cur.whichOutput )
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: odd spacing.
if "idx+1 != partition.ordinality()" is true, you don't need to test "partition.item(idx+1).whichOutput != cur.whichOutput" here, as must be true from line 2907 already.

File: dali/ft/fterror.hpp [hunk 6]
```diff
@@ -147,6 +148,7 @@
 #define DFTERR_WrongRECFMvbBlockSize_Text       "Invalid RECFMvb file Block Size (%d) or the file is not RECFMvb format!"
 #define DFTERR_WrongRECFMvRecordSize_Text       "Invalid RECFMv file Record Size (%d) or the file is not RECFMv format!"
 #define DFTERR_WrongSplitRecordSize_Text        "Invalid Record Size (%d, 0x%08x)!"
+#define DFTERR_InvalidXmlPartSize_Text			"Invalid XML part size:%d! Size is less than XML Header (%d) + Footer (%d)) size!"
```
Comment (jakesmith) [inline dali/ft/fterror.hpp]: Looks like tabs instead of spaces in formatting here.

File: ecl/hthor/hthor.cpp [hunk 7]
```diff
@@ -899,15 +899,14 @@ void CHThorXmlWriteActivity::execute()
...
     OwnedRoxieString suppliedHeader(helper.getHeader());
     if (kind==TAKjsonwrite)
-    {
         buildJsonHeader(header, suppliedHeader, rowTag);
-        headerLength = header.length();
-    }
     else if (suppliedHeader)
         header.set(suppliedHeader);
     else
         header.set("<Dataset>\n");
```
Comment (jakesmith) [inline ecl/hthor/hthor.cpp]: Should use your common default #defines, e.g. DEFAULTXMLHEADER

File: ecl/hthor/hthor.cpp [hunk 8]
```diff
@@ -958,11 +955,20 @@ void CHThorXmlWriteActivity::setFormat(IFileDescriptor * desc)
...
+        if (headerLength)
+            desc->queryProperties().setPropInt("@headerLength", headerLength);
+
+        if (footerLength)
+            desc->queryProperties().setPropInt("@footerLength", footerLength);
+    }
+    else
+    {
+        desc->queryPart(0)->queryProperties().setPropInt("@headerLength", headerLength);
+        desc->queryPart(0)->queryProperties().setPropInt("@footerLength", footerLength);
```
Comment (jakesmith) [inline ecl/hthor/hthor.cpp]: it may make sense, but could do with a comment here to explain why json does not set @headerLength/@footerLength if 0, but xml does.

File: roxie/ccd/ccdserver.cpp [hunk 9]
```diff
@@ -11273,10 +11270,25 @@ class CRoxieServerXmlWriteActivity : public CRoxieServerDiskWriteActivity
...
+        }
+        else
+        {
+            IPartDescriptorIterator *parts = desc->getIterator();
+            if (parts->first())
+            {
+                do {
+                    IPartDescriptor &part = parts->query();
+                    part.queryProperties().setPropInt("@headerLength", headerLength);
+                    part.queryProperties().setPropInt("@footerLength", footerLength);
```
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: Similar to hthor comment:
it may make sense, but could do with a comment here to explain why json does not set @headerLength/@footerLength if 0, but xml does.

Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: But also, why are @headerLength and @footerLength set as File attributes if JSON, but set per Part if XML ?

File: thorlcr/activities/xmlwrite/thxmlwrite.cpp [hunk 10]
```diff
@@ -45,7 +45,7 @@ class CXmlWriteActivityMaster : public CWriteMasterBase
         StringBuffer rowTag;
         OwnedRoxieString xmlpath(helper->getXmlIteratorPath());
         if (!xmlpath)
-            rowTag.append("Row");
+            rowTag.append(DEFAULTXMLROWTAG);
```
Comment (jakesmith) [inline thorlcr/activities/xmlwrite/thxmlwrite.cpp]: Good to use these, but should put them in common place and also use them in hthor + Roxie.

File: thorlcr/activities/xmlwrite/thxmlwrite.cpp [hunk 11]
```diff
@@ -66,6 +66,28 @@ class CXmlWriteActivityMaster : public CWriteMasterBase
...
+                footerLength = supplied.set(DEFAULTXMLFOOTER).newline().length();
+
+            IPartDescriptorIterator *parts = fileDesc->getIterator();
+            if (parts->first())
+            {
+                do {
+                    IPartDescriptor &part = parts->query();
+                    part.queryProperties().setPropInt("@headerLength", headerLength);
+                    part.queryProperties().setPropInt("@footerLength", footerLength);
+                } while (parts->next());
```
Comment (jakesmith) [inline thorlcr/activities/xmlwrite/thxmlwrite.cpp]: Why are these set per part in XML case and at the File level in JSON case?

File: thorlcr/mfilemanager/thmfilemanager.hpp [hunk 12]
```diff
@@ -40,6 +40,10 @@
 
 #define PAUSETMPSCOPE "thorpause"
 
+#define DEFAULTXMLROWTAG "Row"
+#define DEFAULTXMLHEADER "<Dataset>"
+#define DEFAULTXMLFOOTER "</Dataset>"
```
Comment (jakesmith) [inline thorlcr/mfilemanager/thmfilemanager.hpp]: As mentioned in other comment, these would be better place in common file and used by hthor/roxie too.

**Comment by jakesmith** [comment]
@AttilaVamos - look sensible, but have a few questions/comments.

File: ecl/hthor/hthor.cpp (hunk 8 above)
Comment (AttilaVamos) [inline ecl/hthor/hthor.cpp]: I think we should ask @afishbeck about json part.

Comment (afishbeck) [inline ecl/hthor/hthor.cpp]: For JSON if the attribute doesn't exist the default is 0, but you can common it up and always write the value out.

File: roxie/ccd/ccdserver.cpp (hunk 9 above)
Comment (afishbeck) [inline roxie/ccd/ccdserver.cpp]: I thought the header and footer were the same on all parts?  If so it can (like JSON) be kept at the file level?

Comment (AttilaVamos) [inline roxie/ccd/ccdserver.cpp]: Unfortunately it is not. 
In the special case when user wants to spray several "small" XML files (with different [size] header and/or footer – they are 100% valid XML files) into a multi slave Thor system (N:M). Our existing spray stores the header of the first XML file into the first part, then add content for several input  files (and drop out their header and footer) till reach the first part boundary, then gets the next XML file (e.g. 19th) and its header stores into the next part header and so on. And same thing is happening with footers, too.
(I faced this situation is couple of times in past 2 years). 
In this case each part can have a different (size) header/footer. That is why I stored lengths into the parts and use them for despray. 

I can imagine in the near future somebody wants to spray multiple small JSON file with different header/footer and the situation will be same.

Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: >Our existing spray stores the header of the first XML file into the first part, then add content for several input files (and drop out their header and footer) till reach the first part boundary, then gets the next XML file (e.g. 19th) and its header stores into the next part header and so on.

Not sure I understand.. what happens if you spray 4 small XML files (all with different sized headers) to a 2 node system?

Comment (afishbeck) [inline roxie/ccd/ccdserver.cpp]: Are these small files split into multiple even smaller parts?

File: dali/ft/filecopy.cpp (hunk 1 above)
Comment (AttilaVamos) [inline dali/ft/filecopy.cpp]: If there are no xml header or footer then then xmlHeaderLength and xmlFooterLength are 0.

**Comment by AttilaVamos** [comment]
@jakesmith @afishbeck updated, please re-review.

**Comment by AttilaVamos** [comment]
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

**Comment by richardkchapman** [comment]
@jakesmith @afishbeck updated, please re-review.

File: roxie/ccd/ccdserver.cpp (hunk 9 above)
Comment (jakesmith) [inline roxie/ccd/ccdserver.cpp]: I see the added comment:
```
// If multiple XML files with different header/footer sprayed into a
// multi slaves Thor system then every parts can have different length
// header/footer. For compatibility reason Roxie does same.
```

But I'm still dubious this change (setting per Part) is correct here (and in similar change in hthor/thor code).
Spray might be spraying files with different header/footer lengths, but here in the engines, all header/footer lengths must be same for each part.

I see code elsewhere reading headerLength at the File level.
The one place that reads it at the Part level, FileSprayer::analyseFileHeaders, would appear to also read it at the File level and use it as the default (if nothing specified at the part level).

So I would have thought this code in the engines should just set at the File level.

@AttilaVamos - can you explain why not?

File: dali/ft/filecopy.cpp [hunk 13]
```diff
@@ -88,6 +88,8 @@
 #define FAcrc               "@fileCrc"
 #define FAsize              "@size"
 #define FAcompressedSize    "@compressedSize"
+#define FAheaderLength      "@headerLength"
+#define FAfooterLength      "@footerLength"
```
Comment (jakesmith) [inline dali/ft/filecopy.cpp]: Minor: These definitions are duplicates (with subtly difference names (FA* vs FP*) of the definitions in ftbase.cpp. They should be commoned up somewhere and used in-place of the remaining places that use e.g. '@headerLength' directly.

**Comment by jakesmith** [comment]
@afishbeck - could you also look please

File: dali/ft/filecopy.cpp [hunk 14]
```diff
@@ -1501,6 +1503,13 @@ void FileSprayer::analyseFileHeaders(bool setcurheadersize)
         {
             try
             {
+                if (distributedSource)
+                {
+                    Owned<IDistributedFilePart> curPart = distributedSource->getPart(idx);
+                    IPropertyTree& curProps = curPart->queryAttributes();
+                    cur.xmlHeaderLength = curProps.getPropInt(FAheaderLength, 0);
```
Comment (afishbeck) [inline dali/ft/filecopy.cpp]: Are these values immediately being rewritten by either locateContentHeader or srcFormat.headerLength, in the next few lines?

I think what you want would be 1. if the values are in file level DFS use it.  2. or if the values are in PART level DFS, use it.  3. otherwise locateContentHeader.

File: dali/ft/filecopy.cpp [hunk 15]
```diff
@@ -2873,6 +2914,16 @@ void FileSprayer::updateTargetProperties()
                 // TODO: Create DistributedFilePropertyLock for parts
                 curPart->lockProperties();
                 IPropertyTree& curProps = curPart->queryAttributes();
+
+                FilePartInfo & curHeaderSource = sources.item(whichHeaderInput);
+                curProps.setPropInt(FAheaderLength, curHeaderSource.xmlHeaderLength);
```
Comment (afishbeck) [inline dali/ft/filecopy.cpp]: Are these always getting set?  Would it be better to set them at the file level if they are all the same, then set them at the part level only if necessary?

File: ecl/hthor/hthor.cpp [hunk 16]
```diff
@@ -958,11 +955,20 @@ void CHThorXmlWriteActivity::setFormat(IFileDescriptor * desc)
...
         desc->queryProperties().setPropInt("@headerLength", headerLength);
-    if (footerLength)
         desc->queryProperties().setPropInt("@footerLength", footerLength);
+    }
+    else
+    {
+        // If multiple XML files with different header/footer sprayed into a
+        // multi slaves Thor system then every parts can have different length
+        // header/footer. For compatibility reason Hthor does same.
+        desc->queryPart(0)->queryProperties().setPropInt("@headerLength", headerLength);
```
Comment (afishbeck) [inline ecl/hthor/hthor.cpp]: Again if for files for which each part has the same header length you write it to the file properties instead of each part, then this can be the same as json.  Because ECL writes files where the header and footer are all the same.

This makes it clear to me why it should be in the File attributes whenever possible.

File: roxie/ccd/ccdserver.cpp (hunk 9 above)
Comment (afishbeck) [inline roxie/ccd/ccdserver.cpp]: @jakesmith 
@AttilaVamos 
After looking closer I think when ECL writes the file both json and xml should save the header lengths at the FILE, not the PART level.  ECL will never create different headerlengths per part.

When we read in the lengths, we can check the file level first, only if its not there would we check per part.

Also I think I should look at fixing JSON to always write the header lengths, even if 0.

File: thorlcr/activities/xmlwrite/thxmlwrite.cpp (hunk 11 above)
Comment (afishbeck) [inline thorlcr/activities/xmlwrite/thxmlwrite.cpp]: Again, see note above about avoiding per part lengths when not necessary (and when ECL is writing the file it's never necessary).

**Comment by afishbeck** [comment]
@AttilaVamos main gist of my comments are only write per part header lengths when spraying multiple files that cause header lengths to vary.  When spraying a single file, or writing a file from ECL you can save the headerlength at the file level.

When reading the lengths you can check file level first, if not there then check part level.

**Comment by AttilaVamos** [comment]
@jakesmith @afishbeck updated (but not fully tested), please re-review.

File: roxie/ccd/ccdserver.cpp [hunk 17]
```diff
@@ -11224,7 +11225,7 @@ class CRoxieServerXmlWriteActivity : public CRoxieServerDiskWriteActivity
         CRoxieServerDiskWriteActivity::start(parentExtractSize, parentExtract, paused);
         OwnedRoxieString xmlpath(xmlHelper.getXmlIteratorPath());
         if (!xmlpath)
-            rowTag.set("Row");
+            rowTag.set(DEFAULTXMLHEADER);
```
Comment (afishbeck) [inline roxie/ccd/ccdserver.cpp]: There are quite a few places like this where you are using DEFAULTXMLHEADER, but I think you meant DEFAULTXMLROWTAG?

Comment (AttilaVamos) [inline roxie/ccd/ccdserver.cpp]: Yes, you are right. (That could be a kind of midnight mistake.)

**Comment by AttilaVamos** [comment]
@afishbeck  wrong substitutes of "Row" are corrected.

File: dali/ft/filecopy.cpp [hunk 18]
```diff
@@ -1943,9 +1998,11 @@ void FileSprayer::cloneHeaderFooter(unsigned idx, bool isHeader)
     PartitionPoint & next = * new PartitionPoint;
     //NB: headerSize include the size of the xmlHeader; size includes neither header or footers.
     if (isHeader)
+        // Set offset to the XML header
         next.inputOffset = curSrc.headerSize - curSrc.xmlHeaderLength;
     else
-        next.inputOffset = curSrc.headerSize + curSrc.size;
+        //Set offset to the XML footer
+        next.inputOffset = curSrc.size + curSrc.xmlHeaderLength;
```
Comment (afishbeck) [inline dali/ft/filecopy.cpp]: @AttilaVamos I'm curious what made this have to change?  Any chance this could break non-xml formats?

Comment (afishbeck) [inline dali/ft/filecopy.cpp]: Also did you test this with xml files that start with a BOM?  I think curSrc.headerSize includes the BOM but don't think curSrc.xmlHeaderLength does.

Comment (AttilaVamos) [inline dali/ft/filecopy.cpp]: For the first question: It is a really hard to follow code. Somewhere strange things are going. Sometimes the BOM skipped, sometimes not. The original calculation seems good, but caused skewed footers (I traced it), that's why I change it. If you have any non-xml input file candidate or structure in your mind, please attach it to the JIRA and will test/trace it.

For the second question: Yes I tested it with plain ASCII and utf-8 with 3 bytes of BOM. The result is good.  So somewhere any strange magic happen in this abandoned legacy code. 

If you give me a non-xml file and this code break it then I will raise a different JIRA to fix it.

Comment (afishbeck) [inline dali/ft/filecopy.cpp]: Does the test that fails with skewed footers work in current 5.4 branch?  I'm just wondering what changed that makes this change necessary.
@cloLN Are there test spray files of various formats somewhere?

**Comment by AttilaVamos** [comment]
@afishbeck footer offset calculation reverted.

**Comment by afishbeck** [comment]
@richardkchapman looks fine
